File: core/org.eclipse.ptp.proxy.protocol/util/org/eclipse/ptp/proxy/util/VarInt.java
/*******************************************************************************
 * Copyright (c) 2010 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * IBM Corporation - Initial API and implementation
 *******************************************************************************/
package org.eclipse.ptp.proxy.util;

import java.nio.ByteBuffer;

/**
 * Represent the specified integer value in varint format.
 * 
 * The varint format is described at
 * http://code.google.com/apis/protocolbuffers/docs/encoding.html#varints Note
 * that this function is not thread safe.
 * 
 * @since 5.0
 */
public class VarInt {
	private int fValue = 0;
	private boolean fValid = false;
	private ByteBuffer fBytes = null;

	public VarInt(ByteBuffer buffer) {
		int shift = 0;
		while (buffer.hasRemaining()) {
			byte b = buffer.get();
			fValue |= (b & 0x7f) << shift;
			shift += 7;
			if ((b & 0x80) == 0) {
				fValid = true;
				break;
			}
		}
	}

	public VarInt(int value) {
		fValue = value;
		fValid = true;
	}

	/**
	 * Get a byte buffer representation of this varint. Only valid if
	 * {@link #isValid()} is true.
	 * 
	 * @return byte buffer representation of varint
	 */
	public ByteBuffer getBytes() {
		if (fBytes == null && fValid) {
			fBytes = ByteBuffer.allocate(5); // maximum size
			int val = fValue;
			while (val > 0) {
				byte b = (byte) (val & 0x7f);
				val >>= 7;
				if (val > 0) {
					b |= 0x80;
				}
				fBytes.put(b);
				fBytes.limit(fBytes.position()); // set limit to the size of the
													// buffer
			}
		}
		fBytes.rewind();
		return fBytes;
	}

	/**
	 * Get the integer value of this varin. Only valid if {@link #isValid()} is
	 * true.
	 * 
	 * @return integer value of varint
	 */
	public int getValue() {
		return fValue;
	}

	/**
	 * Test if this varint is valid.
	 * 
	 * @return true if the varint is valid, false otherwise
	 */
	public boolean isValid() {
		return fValid;
	}
}

File: core/org.eclipse.ptp.utils/src/varint.c
/*
 * Copyright (c) 2010 IBM Corporation.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

#include <stdio.h>

#include "varint.h"


/*
 * Convert the specified integer value to varint format and copy to the
 * supplied buffer. If 'end' is not NULL, return a pointer to the next
 * character after the varint in the buffer. Returns the length of
 * the converted varint.
 *
 * The varint format is described at
 * http://code.google.com/apis/protocolbuffers/docs/encoding.html#varints
 * Note that this function is not thread safe.
 */
int
varint_encode(int value, unsigned char *buf, unsigned char **end)
{
	unsigned char *	p = buf;
	unsigned int 	num;

	num = (unsigned int) value;
	do {
		*p++ = (num & 0x7f) | 0x80;
		num >>= 7;
	} while (num != 0);

	*(p-1) &= 0x7f;

	if (end != NULL) {
		*end = p;
	}

	return p - buf;
}


/*
 * Convert the data specified by varint_p to integer from varint format and
 * store the value in 'result'. If 'end' is not NULL, store a pointer to the next
 * character after the varint in the buffer. Returns the length of the
 * varint in the buffer.
 */
int
varint_decode(int *result, unsigned char *varint_p, unsigned char **end)
{
	unsigned char *p;
	int shift;
	int length;
	int value;

	p = varint_p;
	length = 0;
	value = 0;
	shift = 0;
	for (;;) {
		value |= ((*p & 0x7f) << shift);
		length++;
		if ((*p & 0x80) == 0x0) {
			break;
		}
		p++;
		shift += 7;
	}

	if (end != NULL) {
		*end = ++p;
	}

	*result = value;

	return length;
}

/*
 * Return the number of bytes required to hold an integer in varint format
 */
int varint_length(int num)
{
	if (num < 0) {
		return 5;
	}
	else {
		if (num < (1 << 7)) {
			return 1;
		}
		else if (num < (1 << 14)) {
			return 2;
		}
		else if (num < (1 << 21)) {
			return 3;
		}
		else if (num < (1 << 28)) {
			return 4;
		}
		else {
			return 5;
		}
	}
}
